### PR TITLE
SILGen support for throwing indirect errors

### PR DESCRIPTION
**Pull request by slavapestov**
In a typed throws world, the thrown error type might be address-only. In this case we return it via the new `@error_indirect` convention. The type lowering and SILGen prolog generation for this was previously landed in https://github.com/apple/swift/pull/69430. This PR makes `throw` and `try call(...)` work.

Throwing is handled with a new `throw_addr` instruction; instead of taking an operand it assumes the indirect error result was initialized already.

If the callee has an indirect error result, `try_apply` takes a temporary allocation to use as the indirect error result. Then, instead of the thrown error being an argument to the error basic block, we instead assume the indirect error result is live in the error block.

Note that `do {} catch {}` doesn't work with indirect errors yet.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: docs/SIL.rst [hunk 1]
```diff
@@ -8006,6 +8006,25 @@ the basic block argument will be the operand of the ``throw``.
...
+``````````
+::
+
+  sil-terminator ::= 'throw_addr'
+
+  throw_addr
+  // indirect error result must be initialized at this point
+
+Exits the current function and returns control to the calling
+function. The current function must have an ndirect error result,
```
Comment (eeckstein) [inline docs/SIL.rst:8019]: ... indirect ...

**Review by eeckstein — APPROVED** [review]
lgtm!

File: docs/SIL.rst [hunk 2]
```diff
@@ -8006,6 +8006,25 @@ the basic block argument will be the operand of the ``throw``.
...
+and so the function must have been invoked with a ``try_apply``
+instruction.  Control will resume in the error destination of
+that instruction.
```
Comment (eeckstein) [inline docs/SIL.rst:8022]: Maybe also add that the function is responsible to initialize the indirect error result before `throw_addr`.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Pull request merged**